// time-number/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.timeFromInt = timeFromInt;
exports.timeToInt = timeToInt;
function isOwnException(ex) {
  return ex.message.substring(0, 11) === 'time-number';
}

function leftPadding(value) {
  var len = arguments.length <= 1 || arguments[1] === undefined ? 2 : arguments[1];

  var sValue = value.toString();
  var lLen = len - sValue.length;

  var l = '';

  while (l.length < lLen) {
    l += '0';
  }

  return '' + l + sValue;
}

function internalTimeFromInt(val, isValidationEnabled) {
  var value = parseInt(val, 10);

  if (isValidationEnabled && (value < 0 || value >= 24 * 3600)) {
    throw new RangeError('time-number, timeFromInt(): rangeError, value supposed to be between 0 and 86399');
  }

  var hours = Math.floor(value / 3600);
  var minutes = Math.floor((value - hours * 3600) / 60);
  var seconds = value - hours * 3600 - minutes * 60;

  var ret = [leftPadding(hours), leftPadding(minutes)];

  if (seconds) {
    ret.push(leftPadding(seconds));
  }

  return ret.join(':');
}

function timeFromInt(val) {
  var isValidationEnabled = arguments.length <= 1 || arguments[1] === undefined ? true : arguments[1];

  if (!isValidationEnabled) {
    return internalTimeFromInt(val, isValidationEnabled);
  }

  try {
    if (val - parseFloat(val, 10) + 1 >= 0) {
      return internalTimeFromInt(val, isValidationEnabled);
    }

    throw new Error();
  } catch (ex) {
    if (isOwnException(ex)) {
      throw ex;
    }

    throw new Error('time-number, timeFromInt(): invalud value: \'' + val + '\', supposed to be number');
  }
};

function internalTimeToInt(val, isValidationEnabled) {
  var times = val.split(':');
  var len = times.length;

  while (times.length < 3) {
    times.push('0');
  }

  var iTimes = times.map(function (i) {
    return parseInt(i, 10);
  });

  if (isValidationEnabled) {
    var hours = iTimes[0];

    if (hours < 0 || hours > 23) {
      throw new RangeError('time-number, timeToInt(): hours must be between 0 and 23, provided value: \'' + val + '\'');
    }

    if (len > 1) {
      var minutes = iTimes[1];

      if (minutes < 0 || minutes > 59) {
        throw new RangeError('time-number, timeToInt(): minutes must be between 0 and 59, provided value: \'' + val + '\'');
      }
    }

    if (len > 2) {
      var seconds = iTimes[2];

      if (seconds < 0 || seconds > 59) {
        throw new RangeError('time-number, timeToInt(): seconds must be between 0 and 59, provided value: \'' + val + '\'');
      }
    }
  }

  return iTimes[0] * 3600 + iTimes[1] * 60 + iTimes[2];
}

function timeToInt(val) {
  var isValidationEnabled = arguments.length <= 1 || arguments[1] === undefined ? true : arguments[1];

  if (!isValidationEnabled) {
    return internalTimeToInt(val, isValidationEnabled);
  }

  try {
    if (!val.match(/^\d+(:\d+(:\d+)?)?$/)) {
      throw new Error();
    }

    return internalTimeToInt(val, isValidationEnabled);
  } catch (ex) {
    if (isOwnException(ex)) {
      throw ex;
    }

    throw new Error('time-number, timeToInt(): supported formats are \'HH\', \'HH:mm\', \'HH:mm:ss\', provided value: \'' + val + '\' doesn\'t match any of them');
  }
};
